### ISSUE TITLE
Do SkuDetails / skuDetailsToken expire?

### ISSUE THREAD

**Issue by arneee** [android/play-billing-samples#440]
Hi,

I could find it anywhere in the docs, but it seems like the queried SkuDetails expire after some time.

My app is a utility app which sometimes runs for hours or even days in the background. I query the SKUs at the app start to be able to show the prices. If the user starts the purchase flow immediately everything works. If he does that after a day, it doesn't anymore  with the error message "5 / Invalid SKU details." The only thing different between a working and not working SkuDetails is the skuDetailsToken.

So please add to the docs if the SkuDetails expire and after what time that is.

**Comment by vitaliyInnowate** [comment]
Hey!

I believe I have the same issue as you when loading SkuDetails from cache and trying to start purchase. It leads to "Invalid SKU details" when old cached token differs from the one we freshly get with query. Also probably it happens only for live app, but that isn't proved 100%. Unfortunately no details in Google's documentation as usual.

**Comment by giapmn-1380** [comment]
Hi, I have the same issue, and I can't find any documents from Google.

**Comment by ducnt-1461** [comment]
i also have same problem when saving sku detail in database, if data is not updated this problem will happen

**Comment by tungnk-2476** [comment]
Hi everybody, I have the same issue. I have tried many ways but not working :(

**Comment by huydd-1086** [comment]
I have same issue. Anyone have solution for this?

**Comment by NguyenTaTrung** [comment]
I also had the same problem, i have been stuck for almost a day to find a solution but still can't work. Anyone have an idea?

**Comment by tgdong2296** [comment]
I have same issue. Anyone have solution ?

**Comment by lienpv-1080** [comment]
I have the same issue. Anyone have solution ?

**Comment by thoalt-1666** [comment]
I have same issue. Anyone have solution for this?

**Comment by hieunv-2463** [comment]
I have the same issue and i need some help !!!

**Comment by stephanedupont** [comment]
@hieunv-2463, @thoalt-1666, @lienpv-1080, @tgdong2296, @NguyenTaTrung, @huydd-1086, @tungnk-2476: if the OP is correct, shouldn't querying the details again from time to time so they don't expire work? Did you try this?

**Comment by giapmn-1380** [comment]
@stephanedupont 
sorry, maybe i don't understand

I querying one time the SkuDetails at the start app, and then i save it to database. When i perform purchase item, i get skuDetail from database and flow immediately everything works.

However, maybe at any time,  When i perform purchase item **launchBillingFlow**, it doesn't anymore with the error message "5 / Invalid SKU details." The only thing different between a working and not working SkuDetails is the skuDetailsToken.

so, data skuDetails from database mayble expire??

My solution:
Don't fetch skuDetails when start app, don't save skuDetails to cache database. SkuDetails will be get at the time buy item.

I have a question?
 skuDetailsToken (or SkuDetails) maybe expire?? When does it expire?

**Comment by stephanedupont** [comment]
@giapmn-1380 Yes, I understand that. What I don't understand is why everyone else asks for a solution when there is a simple solution.

**Comment by giapmn-1380** [comment]
@stephanedupont thanks u so much :D, but i'm still confused :(

> skuDetailsToken (or SkuDetails) maybe expire?? When does it expire?

**Comment by stephanedupont** [comment]
@giapmn-1380 I have no idea. What I know is that I always implemented an expiration time when I cached the sku details, and I never had the problem.

**Comment by vitaliyInnowate** [comment]
> @giapmn-1380 Yes, I understand that. What I don't understand is why everyone else asks for a solution when there is a simple solution.

There is no simple solution as you said because:
1. There is no documentation which will explain this mechanics. Everything we do here is our own fantasies on what token is and how it works.
2. There is no callback or other way to know your token is expired and your next call to launch billing flow will fail. Why should one request new product each 1 hour? Why exactly 1 hour and not 3 hours or 45 minutes? Do all clients have this cap?

It's understandable that in order to minimize risks we should:
1. Request products and rewrite tokens each app start.
2. Do not use cached products to launch billing flow or other operations except visualization.
3. Request products each X hours in order to keep them up-to-date. Let's say this period to be 1 hour, ok.

But it is not a solid engineering solution, it may fail or it may work depending on your luck and some unknown Google's rules.

**Comment by stephanedupont** [comment]
> > @giapmn-1380 Yes, I understand that. What I don't understand is why everyone else asks for a solution when there is a simple solution.
> 
> There is no simple solution as you said because:
> 
> 1. There is no documentation which will explain this mechanics. Everything we do here is our own fantasies on what token is and how it works.
> 2. There is no callback or other way to know your token is expired and your next call to launch billing flow will fail. Why should one request new product each 1 hour? Why exactly 1 hour and not 3 hours or 45 minutes? Do all clients have this cap?
> 
> It's understandable that in order to minimize risks we should:
> 
> 1. Request products and rewrite tokens each app start.
> 2. Do not use cached products to launch billing flow or other operations except visualization.
> 3. Request products each X hours in order to keep them up-to-date. Let's say this period to be 1 hour, ok.
> 
> But it is not a solid engineering solution, it may fail or it may work depending on your luck and some unknown Google's rules.

Well, yes, I agree that this is not a perfect solution because of what you mention. But a lot of things around this library are not perfect and depends on "luck and some unknown Google'es rules" anyway :)

But this is simple and a probably-good-enough solution nevertheless. I have several quite successful apps that implement a one hour cache, and I never encountered this error once. So I think it's safe to assume that one hour is safe to use.

If someone has a different experience with such a value, please share.

**Comment by stephanedupont** [comment]
(but yes, of course, this should be clarified by Google in the documentation!)

**Comment by Lingviston** [comment]
I **always** request fresh `SkuDetails` before launching purchase flow. However, a very small cohort of users faces this issue anyway.

**Comment by algrid** [comment]
I have the same problem, also for a very small subset of users. Getting "Expired Product details. Please fetch product details again and use it to retry the call." and response code: 5. And the problem doesn't go away after product details update.
